074-connector-offsets-support.md

```diff
@@ -0,0 +1,180 @@
+      },
+      "offset": {
+        "kafka_offset": // Kafka offset
+      }
+    }
+  ]
+}
+```
+
+### Listing offsets
+The KafkaConnector resource will be updated with a new property to enable the current offsets to be listed in the resource status.
+
+The new property will be a boolean property called `showOffsets` and default to `false`. It should default to `false` because enabling it requires the operator to do an additional API call every time it reconciles the connector.
+If there are a lot of connectors running in the cluster this is a lot of additional API calls.
+
+When the property is set to `true` the status will include the offsets.
+
+The details in the status will be similar to the format returned by the Connect API:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnector
+#...
+status:
+  #...
+  connectorOffsets:
+    lastUpdateTime: "2024-05-21T09:08:58.467712198Z"
+    source:
+      - partition: "{\"connector-determined-partition\": \"1\"}" # value is a String containing JSON as received from the API
+        offset: "{\"connector-determined-offset\": \"1\"}" # value is a String containing JSON as received from the API
+    sink:
+      - partition:
+          kafkaTopic: "my-topic"
+          kafkaPartition: 1
+        offset:
+          kafkaOffset: 0
+```
+
+Notes:
+* Only one of the `source` and `sink` properties will be present.
+* The operator will update the `lastUpdateTime` whenever it updates the offsets section of the status.
+* If the `showOffsets` property is set to `false` the operator will leave the last observed offsets in the status, keeping the `lastUpdateTime` so the user sees the offsets list has not been updated.
+
+### Altering offsets
+To alter offsets a new `KafkaConnectorOffsets` custom resource will be added.
+This is similar to the approach used for Cruise Control with `KafkaRebalance`.
+This approach is used because the request requires a complex object to specify the desired offsets.
+
+The new resource should look like:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnectorOffsets
+metadata:
+  name: my-source-connector-offsets
+  labels:
+    # The strimzi.io/cluster label identifies the KafkaConnect instance
+    # in which this connector is running
+    strimzi.io/cluster: my-connect-cluster
+spec:
+  connectorName: my-source-connector
+  sourceOffsets:
+    - partition: "{\"connector-determined-partition\": \"1\"}" # value is a String containing JSON can be directly passed to the API
+      offset: "{\"connector-determined-offset\": \"1\"}" # value is a String containing JSON can be directly passed to the API
+  sinkOffsets:
+    - partition:
+        kafkaTopic: "my-topic"
+        kafkaPartition: 1
+      offset:
+        kafkaOffset: 0
+```
+
+Notes:
+* Only one of `sourceOffsets` and `sinkOffsets` should be specified.
+* The format matches the format shown in the status when listing offsets so it can be copy pasted here.
```

Comment:
I don't follow this question, I meant that I have chosen to use the same format in the offsets CR as the status section of the KafkaConnector CR, just to make it easier for people interacting with the CRs

Reply:
So, if I got it right, the offset format is defined by the connector. So what happens if I tell Strimzi to set the offsets to a JSON that is not valid for given connector? Are there any validation facilities in Connect for this? Or will Connect just store it in the topic and break the connector?

The logic that it can be copy-pasted is fine. But you still expect the user to change the current status, so there is a pretty big space to create mistakes.

Reply:
That sounds like a great design :-D. Anyway, not much one can do - maybe we should clarify it in the proposal at least for future reference?